feat: logout other sessions on email change

syedsajjadkazmii opened the pull request:
Logout of other sessions on email change.

syedsajjadkazmii committed:
feat: logout other sessions on email change

syedsajjadkazmii commented on openedx/core/djangoapps/user_authn/views/login.py:
```diff
@@ -313,6 +313,7 @@ def _handle_successful_authentication_and_login(user, request):
     try:
         django_login(request, user)
         request.session.set_expiry(604800 * 4)
+        request.session['email'] = request.user.email
```
Should we `user.email` here instead of `request.user.email`? @zainab-amir

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -768,6 +768,39 @@ def _get_encrypted_request_headers(request):
         return encrypt_for_log(str(request.headers), getattr(settings, 'SAFE_SESSIONS_DEBUG_PUBLIC_KEY', None))
 
 
+class EmailVerificationMiddleware(MiddlewareMixin):
```
What do you think of something like EmailChangeMiddleware or EmailChangeCheckSessionMiddleware? Note the current name makes me think of the email activation, but maybe that’s just me.

syedsajjadkazmii replied:
You are correct! We will update it with a meaningful name.

robrap reviewed (commented):
I haven’t done a detailed review yet, but wanted to pass on some higher level thoughts. Thank you.

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -768,6 +768,39 @@ def _get_encrypted_request_headers(request):
+                # Flush the session and mark cookies for deletion.
+                request.session.flush()
+                request.user = AnonymousUser()
+                _mark_cookie_for_deletion(request)
```
[Somehow I lost a comment that was meant to be a part of the high-level review from yesterday. Here it is again.]
1. Could you add a `set_custom_attribute` for `session_email_match` as True/False? Search for `.. custom_attribute_name` in github for examples of how to annotate.
2. I recommend adding a setting toggle like `ENFORCE_SESSION_EMAIL_MATCH`, even if annotated as temporary, in case we find exceptions that are valid. This is just meant for rollout.
3. Related, you can see some of the exemptions of SafeSessions by searching for uses of [mark_user_change_as_expected](https://github.com/openedx/edx-platform/blob/1d028084bc731a8842e2216fb8a46094574fdeff/openedx/core/djangoapps/safe_sessions/middleware.py#L942C5-L942C33). Also [see notes about some of the exemptions](https://github.com/openedx/edx-platform/blob/1d028084bc731a8842e2216fb8a46094574fdeff/openedx/core/djangoapps/safe_sessions/middleware.py#L672-L682). Additionally, masquerading needed exemptions. This code for email check may not run into the same issues, but a little testing and a safety net toggle won't hurt.

Please confirm what types of manual testing you have also done, including if you test that masquerading is not affected. Thank you.

robrap replied:
Marking as resolved. Thank you.

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -768,6 +768,39 @@ def _get_encrypted_request_headers(request):
+
+    This middleware ensures that the sessions in other browsers
+    are invalidated (and remain validated in current browser)
+    when a user changes their email in one browser.
```
Nit: It took me a moment to understand the parentheses, so I broke it out.
```suggestion
    This middleware ensures that the sessions in other browsers are invalidated 
    when a user changes their email in one browser. The active session in which
    the email change is made will remain valid.
```

robrap commented on openedx/core/djangoapps/user_authn/views/login.py:
```diff
@@ -315,6 +315,9 @@ def _handle_successful_authentication_and_login(user, request):
         log.debug("Setting user session expiry to 4 weeks")
 
+        # Store the user's email for session consistency (used by EmailVerificationMiddleware)
+        request.session['email'] = user.email
```
I'd refactor this into a method like `EmailVerificationMiddleware.register_email_change(user, request)`, or something of this sort. This way `request.session['email']` is not exposed beyond `EmailVerificationMiddleware`, and it would be easy to find these places where it needed to be set.

robrap commented on common/djangoapps/student/signals/receivers.py:
```diff
@@ -105,8 +105,10 @@ def listen_for_verified_name_approved(sender, user_id, profile_name, **kwargs):
-    """ If user has changed their email, update that in email Braze. """
+def _listen_for_user_email_changed(sender, user, request, **kwargs):
+    """ If user has changed their email, update that in session and Braze profile. """
+    request.session['email'] = user.email
```
See refactor comment in openedx/core/djangoapps/user_authn/views/login.py about wrapping this in a method call.

robrap reviewed (commented):
Thanks. Some additional comments.

syedsajjadkazmii commented:
> Related, you can see some of the exemptions of SafeSessions by searching for uses of [mark_user_change_as_expected](https://github.com/openedx/edx-platform/blob/1d028084bc731a8842e2216fb8a46094574fdeff/openedx/core/djangoapps/safe_sessions/middleware.py#L942C5-L942C33). Also [see notes about some of the exemptions](https://github.com/openedx/edx-platform/blob/1d028084bc731a8842e2216fb8a46094574fdeff/openedx/core/djangoapps/safe_sessions/middleware.py#L672-L682). Additionally, masquerading needed exemptions. This code for email check may not run into the same issues, but a little testing and a safety net toggle won't hurt.
Please confirm what types of manual testing you have also done, including if you test that masquerading is not affected. Thank you.

I manually tested the following scenarios
- With the `ENFORCE_SESSION_EMAIL_MATCH` flag set to false, I logged into both Chrome and Safari using the same email. After changing the email in Safari, I observed that I remained logged in to both browsers.
- With the `ENFORCE_SESSION_EMAIL_MATCH` flag set to true, I again logged into both Chrome and Safari with the same email. Following an email change in Safari, I observed that I remained logged in to Safari, but the session was invalidated in Chrome.

Regarding masquerading, I interpreted it as the ability to view the platform as other users in the dashboard. I tested it by masquerading as other users in the dashboard, and everything worked as expected. However, if there's a different interpretation or specific aspects related to masquerading that require testing, please provide clarification. I would be happy to test that too.

robrap commented:
> Regarding masquerading, I interpreted it as the ability to view the platform as other users in the dashboard. I tested it by masquerading as other users in the dashboard, and everything worked as expected. However, if there's a different interpretation or specific aspects related to masquerading that require testing, please provide clarification. I would be happy to test that too.

Yes. Just making sure masquerading still works appropriately when `ENFORCE_SESSION_EMAIL_MATCH` is enabled. I think masquerading messes with the session, but maybe it does so outside of the scope of your check, or maybe not in the way I think. So, as long as it isn't broken, that sounds good.

robrap commented on cms/djangoapps/contentstore/tests/tests.py:
```diff
@@ -85,6 +85,7 @@ def activate_user(self, email):
 
 
 @ddt
+@override_settings(ENFORCE_SESSION_EMAIL_MATCH=False)
```
Note that this is a temporary setting, so if this needs to be disabled to make the tests work, then these tests will need to be updated once the toggle is removed. Consider adjusting or adding individual tests now as-needed, rather than changing for the entire test case.

syedsajjadkazmii replied:
There was no need for this after removing the flag check from login.py file.

robrap commented on common/djangoapps/student/views/management.py:
```diff
@@ -910,7 +911,11 @@ def confirm_email_change(request, key):
-        USER_EMAIL_CHANGED.send(sender=None, user=user)
+        if settings.ENFORCE_SESSION_EMAIL_MATCH:
+            # Update the JWT cookies with new user email
+            response = update_jwt_cookies_in_response(request, response, user)
```
Would it be possible for this to be a part of the `USER_EMAIL_CHANGED` receiver code so we know any email change would get this? Would it be too late for the middleware to take care of this in process_response if register_email_change was called? Something like that? I haven't thought this through, but hopefully you can.

syedsajjadkazmii replied:
I marked a variable `email_changed` to True in the `register_email_change` function and moved this logic in `process_response` of the middleware behind a check if `email_changed` is marked true. Could you please give your opinion on this?

robrap replied:
Thanks. I'm going to add comments on that code.

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -768,6 +768,59 @@ def _get_encrypted_request_headers(request):
+                # Flush the session and mark cookies for deletion.
+                request.session.flush()
+                request.user = AnonymousUser()
+                _mark_cookie_for_deletion(request)
```
A different version of this might be to set variable about whether or not we've matched, and have one custom attribute for whether or not we've matched, and another for the setting `ENFORCE_SESSION_EMAIL_MATCH`. Then flush if there is a mismatch and we are enforcing. In that way, you can observe before you enforce how many requests would be affected, in case there is an issue before you start enforcing. Something to consider.

syedsajjadkazmii replied:
I am sorry but i dont think i was fully able to understand this suggestion here. But I have tried to implement it as much I could understand. Could you please check and let me know if there is anything left from this?

robrap replied:
This is close. To complete what I meant (if you wish), you could rename `_set_session_email_match_custom_attribute` to `_set_session_email_match_custom_attributes`, and inside that method, you could add:
```
set_custom_attribute('is_enforce_session_email_match_enabled', settings.ENFORCE_SESSION_EMAIL_MATCH)
```
From in New Relic, this would just help you see which transactions were from before and after this was enabled.

robrap commented on openedx/core/djangoapps/safe_sessions/tests/test_middleware.py:
```diff
@@ -615,3 +616,126 @@ def test_user_change_with_no_ids(self):
+        # Log in the user
+        self.client.login(email=self.user.email, password=self.TEST_PASSWORD)
+        self.request.session = self.client.session
+        self.request.session['email'] = self.user.email  # Set the session email to match request.user.email
```
Is it possible to have less mocking and more of an integration test by doing something like the earlier test?
```
# simulating email change
user.email = 'new_email@test.com'
user.save()
```

The less mocking the better for this security test. Or maybe there is a separate integration test needed. Anyway, something to consider.

syedsajjadkazmii replied:
I am planning to revise all the tests once we finalize the implementation. So I will surely consider it later.

robrap replied:
I'm going to mark this resolved.

robrap commented on openedx/core/djangoapps/user_authn/cookies.py:
```diff
@@ -180,6 +180,15 @@ def get_response_with_refreshed_jwt_cookies(request, user):
     return response
 
 
+def update_jwt_cookies_in_response(request, response, user):
```
Are you sure we don't want to just use `set_logged_in_cookies` to minimize code and potential sync issues between these cookies?

syedsajjadkazmii replied:
Thanks for poininting it out. You are right about the sync issues. I am now using `set_logged_in_cookies` function.

robrap commented on openedx/core/djangoapps/user_authn/views/login.py:
```diff
@@ -315,6 +318,10 @@ def _handle_successful_authentication_and_login(user, request):
         request.session.set_expiry(604800 * 4)
         log.debug("Setting user session expiry to 4 weeks")
 
+        if settings.ENFORCE_SESSION_EMAIL_MATCH:
```
I wonder if we can drop all of these conditions, and just have the condition protect the flushing of the session? It would be a little riskier, because we'd still be messing with code that could break. But, it would allow observability into what would be enforced before we flip the switch to actually start killing sessions. Something to consider. This would apply to any of these conditionals that come before the final enforcement (and the cookie updates).

syedsajjadkazmii replied:
I have removed the `ENFORCE_SESSION_EMAIL_MATCH` check from the login and email change receiver and moved the cookie update logic to the process_response of middleware. 

I still have the flag check in the `process_response` before updating cookies because if I remove the check from here it will always update the cookies on email change even if the flag is disabled. I am not sure if this is actually a problem or should I remove the flag check from process_response too? Could you please provide your suggestion on this?

robrap replied:
Yes. I missed that there were two things that needed to be protected, the flushing and the updating of the cookies. Looks good.

robrap reviewed (commented):
I added a variety of thoughts. Good luck. You are welcome to reach out to others if you want to make steady progress before January.

syedsajjadkazmii committed:
fix: updated the approach for session invalidation

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -768,6 +768,87 @@ def _get_encrypted_request_headers(request):
+        Sets a variable in the given request object to designate that the email
+        is changed. It will be used in process_response for updating the cookies
+        """
+        request.email_changed = True
```
* [nit] This is just an opinion, but I'd drop this method and just move this to `register_email_change`. It just made me hunt to see that that was the only caller.
* [nit] We usually use the RequestCache for data like this. Note that if we use the RequestCache, it needs to be cleared before each test. You can see examples in other test files that use this.

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -768,6 +768,87 @@ def _get_encrypted_request_headers(request):
+            and self._is_email_change_marked(request)
+        ):
+            # Update the JWT cookies with new user email
+            response = set_logged_in_cookies(request, response, request.user)
```
I've got a number of thoughts related to this. I was confused, and it took me a bit to clarify thing for my self.
* [confirmation requested] Is this a bug fix? In the past, did we simply not update the cookies to match the email change when it was first changed?
* [request] I think we should rename `EmailChangeSessionInvalidationMiddleware` to `EmailChangeMiddleware`, and document that it has two jobs: 1) Upon email change, it will update the session's email and update the JWT cookie to match the new email, and 2) Will invalidate any future session where the user's email does not match its session email.
* [nit] Since `email_changed` (the name you are using for the request flag) could be misinterpreted as a mismatch (e.g. change detected), maybe update this to `email_change_requested` to better clarify that it is only used during a session in which the email change was being made?
* [nit] Lastly, I don't think this needs to be protected by `ENFORCE_SESSION_EMAIL_MATCH`, but I'll leave that up to you. Worst case this would be extra processing if we had a bug, but this isn't as dangerous and clearing the session.

syedsajjadkazmii replied:
Thankyou for the thoughts and suggestions.

> [confirmation requested] Is this a bug fix? In the past, did we simply not update the cookies to match the email change when it was first changed?

We had the cookie update logic in the `confirm_email_change` function which is now moved to `process_response` function of middleware as per the suggestion in this [comment](https://github.com/openedx/edx-platform/pull/33846#discussion_r1433236315). Plus the custom `update_jwt_cookies_in_response` function was removed and we used `set_logged_in_cookies` instead to avoid possible sync issues. 

When the email is changed, the email change will be registered and the `email_change_requested` variable will be set in the request cache. Then the `process_response` function of middleware will make sure that the updated logged-in cookies get set in the response of the update email call.

This was to make sure that JWT authentication also fails on email change.

Please let me know if this is what you were asking.

robrap replied:
This was mostly what I was asking about, but I wasn't asking about an earlier version of this PR, but about the code on master. I was confirming that this fixes a bug that exists on master, where the JWT cookie is not currently updated upon email change. Is that correct? (Note: this isn't a big deal, but it was a part of me understanding the two purposes of the middleware that we documented.)

syedsajjadkazmii replied:
Yes, you are right. We were not updating the JWT cookies previously. So yes, this fixes that too.

robrap reviewed (commented):
Thanks for the updates. Looks good. Just minor requests and thoughts. Thanks.

syedsajjadkazmii commented:
> Thanks for the updates. Looks good. Just minor requests and thoughts. Thanks.

Thanks for the valuable suggestions. I implemented them in the code. Please have a look.

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -768,6 +768,78 @@ def _get_encrypted_request_headers(request):
+        that session and user email change was requested
+        """
+        request.session['email'] = email
+        request_cache.set('email_changed_requested', True)
```
[nit] `email_changed_requested` => `email_change_requested` (all occurrences)

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -768,6 +768,78 @@ def _get_encrypted_request_headers(request):
+    def register_email_change(request, email):
+        """
+        Sets email in session for comparison and sets a variable to mark
+        that session and user email change was requested
```
[Nit] The first line is supposed to be a single line. Something like:
```suggestion
    Stores the fact that an email change happened.

    1. Sets the email in session for later comparison.
    2. Sets a request level variable to mark that the user email change was requested.
```

robrap commented on lms/envs/common.py:
```diff
@@ -5041,6 +5044,19 @@ def _make_locale_paths(settings):  # pylint: disable=missing-function-docstring
+# .. toggle_implementation: DjangoSetting
+# .. toggle_default: False
+# .. toggle_description: When enabled, this setting invalidates sessions in other browsers upon email change,
+#    while preserving the session validity in the browser where the email change occurs.
```
Just adding this note to make the temporary status more clear.
```suggestion
#    while preserving the session validity in the browser where the email change occurs. This toggle
#    is just being used for rollout.
```

robrap reviewed (commented):
Looks good. Thanks. Minor comments.

syedsajjadkazmii committed:
fix: update and add tests

syedsajjadkazmii commented:
Hi @robrap,
I have revised the tests for middleware. Please review them.

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -768,6 +768,80 @@ def _get_encrypted_request_headers(request):
+        # .. custom_attribute_name: session_email_match
+        # .. custom_attribute_description: Indicates whether there is a match between the
+        #      email in the user's session and the current user's email in the request.
+        set_custom_attribute('session_email_match', not are_emails_mismatched)
```
1. Do you have a ticket for removing `ENFORCE_SESSION_EMAIL_MATCH` after rollout?
2. We may want to remove both these custom attributes after rollout, unless we only set this attribute for mismatches. 
3. Also, this could be `set_custom_attribute('session_email_mismatch', are_emails_mismatched)`. After rollout, we could move it under the `if are_emails_mismatched:` conditional.

syedsajjadkazmii replied:
> Do you have a ticket for removing ENFORCE_SESSION_EMAIL_MATCH after rollout?

We will create ticket in our backlog for this

> We may want to remove both these custom attributes after rollout, unless we only set this attribute for mismatches.
> Also, this could be set_custom_attribute('session_email_mismatch', are_emails_mismatched). After rollout, we could move it under the if are_emails_mismatched: conditional.

Yes, we can have this attribute under the  ` if are_emails_mismatched: ` conditional after rollout.

robrap replied:
> Yes, we can have this attribute under the if are_emails_mismatched: conditional after rollout.
Maybe add this to the ticket.

Also, if you haven't already, maybe add a reference to the ticket in the toggle annotation.

robrap replied:
Marking as resolved.

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -768,6 +768,80 @@ def _get_encrypted_request_headers(request):
+            are_emails_mismatched = user_session_email is not None and request.user.email != user_session_email
+            EmailChangeMiddleware._set_session_email_match_custom_attributes(are_emails_mismatched)
+            if settings.ENFORCE_SESSION_EMAIL_MATCH and are_emails_mismatched:
+                # Flush the session and mark cookies for deletion.
```
[nit] I tend to rely on New Relic for observability, but you may want to consider an info log here. Hopefully it should be pretty rare.

robrap commented on openedx/core/djangoapps/safe_sessions/tests/test_middleware.py:
```diff
@@ -615,3 +617,216 @@ def test_user_change_with_no_ids(self):
+    @override_settings(ENFORCE_SESSION_EMAIL_MATCH=True)
+    @patch('openedx.core.djangoapps.safe_sessions.middleware._mark_cookie_for_deletion')
+    @patch("openedx.core.djangoapps.safe_sessions.middleware.set_custom_attribute")
+    def test_process_request_emails_match(
```
Maybe in addition to these tests, but I'm wondering if we could have some more resilient integration test, using the actual middleware, and possible an API call to actually update the user's email?
- In place of calls to `register_email_change`, `process_request` and `process_response`, you could instead make various requests after having set `MIDDLEWARE` appropriately, and ensure that the actual cookies have been deleted or updated, and the user has been logged out, etc.

See https://github.com/openedx/edx-drf-extensions/blob/9eea525bf25a5d2af2911a5e94f3026da649f9e8/edx_rest_framework_extensions/auth/jwt/tests/test_middleware.py#L523-L529 for the type of test I mean, although you probably won't need `ROOT_URLCONF` because you may be using real platform endpoints.

Since this is for security purposes, it would be great to know that all the pieces are working together. That said, I'd also timebox any effort to get this working. Hopefully it would be relatively simple.

syedsajjadkazmii replied:
I am working on this one. Meanwhile please have a look the responses of other comments and let me know about your opinions on them.

robrap replied:
Thanks. As I said - this can be quite useful - but it also can be tricky. If you get stuck, let me know, and maybe I can help or maybe we'll just abandon the idea.

syedsajjadkazmii replied:
@robrap Thankyou. I will first give it a try and let you know.

robrap replied:
Marking this resolved.

robrap commented on openedx/core/djangoapps/safe_sessions/tests/test_middleware.py:
```diff
@@ -615,3 +617,216 @@ def test_user_change_with_no_ids(self):
+        self.request.session = self.client.session
+        self.client.response.set_cookie(settings.SESSION_COOKIE_NAME, 'authenticated')  # Add some logged-in cookie
+
+        # Registering email change (simulating email change requested)
```
1. First, does my suggestion make sense? Is this the difference between this call and the `# simulating email changed` below? Maybe that should rad
```suggestion
        # Simulates original email update (e.g. during account registration, or some past email change).
```
2. Are we missing a test for where there is no previous history of an email change (e.g. all existing sessions that predate this code)? We want to make sure we don't accidentally kill these sessions.

syedsajjadkazmii replied:
> First, does my suggestion make sense? Is this the difference between this call and the # simulating email changed below? Maybe that should rad

The `register_email_change` function has two purposes.

Purpose 1: Stores user email in session on registration, login, and email change to reflect current email of user in session so that it can be used for comparison by the `process_request` function of middleware. 

Purpose 2: Sets a variable in the request cache to indicate that the email was changed so we have to update logged-in cookies with the new email.

Purpose 2 is only required upon successful email change and NOT when the user registers or logs in.

In this test we are only utilizing purpose 1, storing the current email session so that it can be compared in process_request later. 

So I am not sure if we can say `Simulates original email update`? as it's not an email update here actually.

syedsajjadkazmii replied:
> Are we missing a test for where there is no previous history of an email change (e.g. all existing sessions that predate this code)? We want to make sure we don't accidentally kill these sessions.

I have added a test `test_process_request_no_email_change_history` for this purpose. Please have a look at that test.

robrap replied:
> Purpose 1: Stores user email in session on registration, login, and email change to reflect current email of user in session so that it can be used for comparison by the process_request function of middleware.

Should this just be "registration" and "email change"? It seems like "login" would typically not update anything, but instead should fail if mismatched, right? Maybe there is the exceptional case of "login" when session email has never been set before?

Note: I am writing these review comments without studying the code or seeing any of your latest changes. If these comments are not helpful, I plan to do a more complete review at some point.

> So I am not sure if we can say `Simulates original email update`? as it's not an email update here actually.

I meant something more like: `Simulates a past email change (e.g. from an earlier request)`.

Also, if instead of calling register the test first made a "registration" or (second test: "email change") before the request under test, it would ensure everything is hanging together appropriately in all cases. However, this might be the more complete integration test, and here we may just want the comment to clarify where this would have happened in a real situation.

> I have added a test `test_process_request_no_email_change_history` for this purpose. Please have a look at that test.

Thank you. I still need to get to the actual review.

robrap replied:
Marking this as resolved.

robrap commented on openedx/core/djangoapps/safe_sessions/tests/test_middleware.py:
```diff
@@ -615,3 +617,216 @@ def test_user_change_with_no_ids(self):
+        self.client.response.set_cookie(settings.SESSION_COOKIE_NAME, 'authenticated')  # Add some logged-in cookie
+
+        # Registering email change (simulating email change requested)
+        EmailChangeMiddleware.register_email_change(request=self.request, email=self.user.email)
```
1. Could this use the following instead (copied from another test, with updated comment)?
```
        # simulating new email change
        self.user.email = 'new_email@test.com'
        self.user.save()
```
2. I think any test call to `register_email_change` that makes sense to replace with actual user save code (like above), will make a more resilient test.

syedsajjadkazmii replied:
> Could this use the following instead (copied from another test, with updated comment)?

The `process_response` function of middleware relies on the `email_change_requested` variable set in the request cache and it is set in the `register_email_change` function of middleware.

Therefore, `register_email_change` should be called here to set the `email_change_requested` variable. 

But we can make the following change by having both the email change simulation and email change registration. It can make more sense?
```
        # simulating new email change
        self.user.email = 'new_email@test.com'
        self.user.save()
        # Registering email change (simulating email change requested)
        EmailChangeMiddleware.register_email_change(request=self.request, email=self.user.email)
```

Please let me know your thoughts about this.

-------------------------------------------------------------------------------------------------

> I think any test call to register_email_change that makes sense to replace with actual user save code (like above), will make a more resilient test.

There were other uses of `register_email_change` in `test_process_request_emails_match` and `test_process_request_emails_mismatch` tests.  There, the `register_email_change` function was being used for setting user email in session for later comparison by middleware's `proces_request` function. For now I have kept them but we can also remove them and can set the email in session ourselves. as below

```
self.request.session['email'] = self.user.email  # setting email in session for comparison
```

Please let me know your thoughts on this.

robrap replied:
From above:    
```
        # simulating new email change
        self.user.email = 'new_email@test.com'
        self.user.save()
        # Registering email change (simulating email change requested)
        EmailChangeMiddleware.register_email_change(request=self.request, email=self.user.email)
```
> Please let me know your thoughts about this.

The email save doesn't trigger the signal receiver that already calls `register_email_change`?

> There, the `register_email_change` function was being used for setting user email in session for later comparison by middleware's `process_request `function.

Is this where the comment `Simulates a past email change (e.g. from an earlier request)` (or something like that) would apply? If so, I get why it is needed and I'd just like an appropriate comment to clarify why it was needed.

syedsajjadkazmii replied:
> The email save doesn't trigger the signal receiver that already calls register_email_change?

Yes this happens in real situation but the signal receiver is not called with `self.user.save()` so we have to call `register_email_change` explicitly in tests.
But I am trying to write integration tests and hopefully that flow will execute in those tests.

robrap replied:
Marking resolved.

robrap commented on openedx/core/djangoapps/user_authn/views/tests/test_login.py:
```diff
@@ -1179,29 +1179,18 @@ def test_invalid_credentials(self):
-        response = self.client.post(url, {})
-        self.assertHttpBadRequest(response)
+        stored_email = self.client.session.get('email')
+        assert stored_email == self.EMAIL
```
1. This is the important part of this test, but it is out of context of the middleware. I wonder if this test could be removed in favor of the integration test I described in the middleware, which might call this endpoint to update the email?
2. Do we need to separately call/test with the registration endpoint?

syedsajjadkazmii replied:
> Do we need to separately call/test with the registration endpoint?

Thanks. This was an important part we were missing. We were not setting user email in session on registration.

robrap replied:
See other comment about potential alternative solution. Either way, I am marking this resolved. Thanks.

robrap commented on openedx/core/djangoapps/user_authn/views/tests/test_login.py:
```diff
@@ -1179,29 +1179,18 @@ def test_invalid_credentials(self):
         })
         self.assertHttpBadRequest(response)
 
-        # Invalid email address
```
We're all of these test deletions accidental?

syedsajjadkazmii replied:
yes, these were accidental. Thanks for noticing. I reverted them.

robrap reviewed (commented):
Thank you.

syedsajjadkazmii commented:
@robrap Apologies for the delayed response on this. I was on PTO.

syedsajjadkazmii committed:
fix: update tests with descriptive comments

syedsajjadkazmii committed:
feat: add integration tests

syedsajjadkazmii commented:
Hi @robrap,

I have updated the tests' descriptions and tried adding some integration tests. Please have a look at them and provide your suggestions.

I am also facing an issue. As I am using real LMS endpoints, my tests fail in the CMS environment because they could not find that endpoint in CMS. For now, I have used `@skip_unless_lms` but I would need your opinion on this. this is happening for both `reverse('dashboard')` and `reverse('confirm_email_change')`.

robrap commented on lms/envs/common.py:
```diff
@@ -5041,6 +5044,21 @@ def _make_locale_paths(settings):  # pylint: disable=missing-function-docstring
+#       This toggle will be removed in https://2u-internal.atlassian.net/browse/VAN-1797
+# .. toggle_use_cases: temporary
+# .. toggle_creation_date: 2023-12-07
+# .. toggle_target_removal_date: None
```
This isn't a commitment, but it is useful to have a date. Could you make this: 2024-04-01 or 2024-05-01, or whatever date makes sense to you?

robrap commented on openedx/core/djangoapps/safe_sessions/tests/test_middleware.py:
```diff
@@ -615,3 +622,544 @@ def test_user_change_with_no_ids(self):
+        mock_mark_cookie_for_deletion.assert_not_called()
+
+    @patch("openedx.core.djangoapps.safe_sessions.middleware.set_logged_in_cookies")
+    def test_process_response_user_not_authenticated(self, mock_set_logged_in_cookies):
```
It's difficult to tell if this was meant to include `@override_settings(ENFORCE_SESSION_EMAIL_MATCH=True)` or not. Same with some other tests that seem like they _may_ be for `True`, but don't have this set.

An alternative could be to do what we did here with [ForgivingJwtAuthenticationTests](https://github.com/openedx/edx-drf-extensions/blob/9eea525bf25a5d2af2911a5e94f3026da649f9e8/edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py#L646-L650). Each test had conditional assertions for when the toggle was enabled vs disabled.

Note: 
- You have spent a lot of time on this PR, and I know we need to move on. I am **not** saying that this adjustment **must** be done.
- However, there are a lot of tests and it is easy for me to lose track of what cases are covered toggled on vs off. I do wonder if this change might make it easier to follow.

syedsajjadkazmii replied:
> It's difficult to tell if this was meant to include @override_settings(ENFORCE_SESSION_EMAIL_MATCH=True) or not.

About this test: Looking at the [code](https://github.com/openedx/edx-platform/pull/33846/files#diff-3af708aae03ff6e3deae4fd21313de4d139ba95325e4b42c7505928554157cfdR797-R801) of `EmailChangeMiddleware`, the value of `ENFORCE_SESSION_EMAIL_MATCH` toggle doesn't really matter if the user is unauthenticated. Therefore I didn't include a tests for both `True` and `False`. 

> Same with some other tests that seem like they may be for True, but don't have this set.

Your suggestion makes sense, so i will be including tests for both toggle enabled and disabled wherever makes sense. Thankyou

robrap replied:
Please mark this resolved when you are ready. At this point, I think we've spent enough time on tests. After we delete the toggle, the number of tests will be more manageable. Thank you.

robrap commented on openedx/core/djangoapps/safe_sessions/tests/test_middleware.py:
```diff
@@ -615,3 +622,544 @@ def test_user_change_with_no_ids(self):
+        """
+        Integration Test: test that a user remains authenticated upon email change
+        in other browser when ENFORCE_SESSION_EMAIL_MATCH toggle is disabled
+        Verify that the session and cookies are not affected
```
1. I thought the cookies will be updated after email change regardless of the toggle setting? Am I incorrect, or should this also be verified?
2. [Nit] I'm not sure how complicated it would be to instantiate a second Django client to represent the other browser, and actually make requests, rather than dropping back to mocking again. Feel free **not** to not pursue this, especially given the time invested. However, if pursued, maybe a longer test could show the implications in both clients with some test combining.

syedsajjadkazmii replied:
> I thought the cookies will be updated after email change regardless of the toggle setting? Am I incorrect, or should this also be verified?

Yes, you are right. The cookies get updated in the browser where the email change happened. Not in other browsers where the user may have sessions. This test verifies that when the toggle is **disabled** and email is changed in some other browser, the cookies of current browser are not affected (not cleared) and the user remains authenticated in the current browser. On the other case, if the toggle was enabled and email was changed in some other browser, the user would have been logged out of current browser. 

Although its true that the cookies of browser in which the email change has happened, will be updated regardless of toggle setting.

I think I may need to update the description of test for better clarification.

syedsajjadkazmii replied:
> I'm not sure how complicated it would be to instantiate a second Django client to represent the other browser, and actually make requests, rather than dropping back to mocking again. Feel free not to not pursue this, especially given the time invested. However, if pursued, maybe a longer test could show the implications in both clients with some test combining.

For now, we won't be working on this one considering the time invested instead we have created a ticket in our backlog to work on this improvement in tests: https://2u-internal.atlassian.net/browse/VAN-1799. We will surely pick it up and work on it. Feel free to let me know if you want to include any other things in this ticket.

robrap replied:
I think you should forget about this. Tests are good. I added a comment to the ticket. I'm going to mark this resolved as well.

robrap reviewed (commented):
Thank you. I'm going to see what you decide about some of my comments before trying to see if we have all the test cases covered.

robrap commented on openedx/core/djangoapps/safe_sessions/tests/test_middleware.py:
```diff
@@ -615,3 +622,725 @@ def test_user_change_with_no_ids(self):
+        """
+        Calls EmailChangeMiddleware.process_request when there is no previous
+        history of an email change and ENFORCE_SESSION_EMAIL_MATCH is enabled
+        Verifies that existing sessions are not affected.
```
[question] Would it makes sense that at the end of process_response, if the session email is still not yet set, we set it to the current email? This _seems_ like it would be safe, and would allow this to work for existing sessions as well. It would also be good to add something like `set_custom_attribute('session_with_no_email_found', True)` (only in the True case).  This way, no matter what you decide, you should be able to watch and hopefully see this come down to zero. 
- [food for thought] Could this even be an alternative solution to doing anything special in the registration code? I haven't thought this through, but the above might just be a better catch-all, and you could remove the code from registration.

syedsajjadkazmii replied:
It is a very suitable suggestion and yes its an alternative to what we are doing in registration/login.

But there is an issue with this, by this change we get 2 tests failing. ([test_too_many_courses](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/course_api/tests/test_views.py#L405) and [test_get_unknown_course](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/course_home_api/course_metadata/tests/test_views.py#L88).
I was not able to fully understand the reasoning for them but `test_too_many_courses` gives an error `executed 53 queries instead of 50` and my debugging shows that the additional 3 queries executed were of session updation. So it seems to be that having `request.session['email'] = request.user.email` in the middleware is adding extra queries and failing those tests? [I have implemented this solution in my PR and these tests are failing]

Could you please guide on how to resolve this? I would appreciate your opinion/suggestion on this.

robrap replied:
1. I think it is ok to simply adjust 50 to 53 for the first test if we know why it is changing. The query assertions are there to ensure that the number of queries doesn't grow without explanation, or grow very large.
2. Do you understand why 404 => 400? That seems a little more scary to me without understanding why that is changing.

syedsajjadkazmii replied:
> Do you understand why 404 => 400? That seems a little more scary to me without understanding why that is changing.

There is an exception being raised on `request.session.save()` inside `process_response` of django's `SessionMiddleware`  which is causing this 400. This only happens when we set email in session (`request.session['email'] = request.user.email`) in middleware. I am still unable to get an answer on why `request.session.save()` is failing.

robrap replied:
If the new code works for registration, and this might just be a strange case in tests, might it make sense to wrap, swallow, and log/set_custom_attribute for any exception happening for the save, and then continue on rather than raising the exception?

**Update:** We could comment that we think this just happens in tests, unless we see it in production, in which case we can look further. Just an idea.

syedsajjadkazmii replied:
> might it make sense to wrap, swallow, and log/set_custom_attribute for any exception happening for the save, and then continue on rather than raising the exception?

Unfortunately, I don't think we can wrap the `request.session.save()` for exception since its inside the `SessionMiddleware.process_response` which is inherited by `SafeSessionMiddleware`. 

Some further debugging shows that its probably due to the `raise Http404()`.  `request.session.save()` fails if view calls `raise Http404()` and `request.session` is updated (e.g, `request.session['email'] = request.user.email`).  It works fine if the view returns normal `HttpResponse`. 
I also tried reproducing the test's flow locally but it was working as expected (returning 404 instead of 400). I am still unclear on the reason why `request.session.save()` getting failed in the test but i am trying to find the reason.

Note: I am just giving an update on my current findings.

syedsajjadkazmii replied:
Further findings

The Django test cases, specifically the test case in question, are [wrapped inside](https://docs.djangoproject.com/en/5.0/topics/db/transactions/#use-in-tests) `transaction.atomic()`. When an error occurs within the transaction (such as raising `Http404()` in our case), it leads to the failure of subsequent database operations (e.g., `request.session.save()`). This is confirmed by the following error in stack trace of test failure.

```
raise TransactionManagementError(
django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/sessions/middleware.py", line 61, in process_response
    request.session.save()
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/sessions/backends/db.py", line 94, in save
    raise UpdateError
django.contrib.sessions.backends.base.UpdateError
```

Now this answers the questions I guess

**Why was it happening for test only not in real execution?**
Because only tests are wrapped inside `transaction.atomic()` to make sure that db transactions can be rolled back at the end of tests.

**Why did it occur for `Http404` but not for normal `HttpResponseForbidden`?**
`Http404()` is an exception, causing `transaction.atomic()` to fail, while `HttpResponseForbidden` is a regular response with a 404 status code.

**Why were there database transactions in tests?**
Django automatically switches to a database session store for tests, ensuring an isolated and consistent testing environment.

**Why didn't it occur for successful responses?**
Success responses were instances of `HttpResponse` and not exceptions thus not causing the transaction failure.

syedsajjadkazmii replied:
Based on this, i think we would need to move the `request.session['email'] = request.user.email` back to login/register views. (mean move `register_email_change` back to login/register views) What do you think?

robrap replied:
As discussed in our meeting, we think we have the right solution and you will experiment with finding the best way to adjust tests. We'll see what you find.

syedsajjadkazmii replied:
My findings:

[test_get_unknown_course](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/course_home_api/course_metadata/tests/test_views.py#L88) was the only test having the `TransactionManagement` error. Other tests (around 4 to 5 in count) only needed query count adjustments. 

Currently, I have used the approach of having the settings flag for that test (as we discussed). But as `test_get_unknown_course` is the only test getting affected by the `TransactionManagement` error. 
I was wondering if we could simply add `self.client.logout()` in that test and make that work? That test is just checking/testing that the view returns `404` for an invalid course ID. I have verified that the view returns `404` for invalid course IDs for both authenticated and unauthenticated users. 

This way we wouldn't have to define a settings flag just for one test (that too isn't majority concerned with user authentication and all the session saving)

Example View URL (with invalid course id):  https://courses.edx.org/api/course_home/course_metadata/course-v1:HarvardX+CS50+X --- you can try hitting this without having a logged-in user.

What are your thoughts on this?

robrap replied:
I think that sounds fine. You could add a brief comment about this, or not if you think it would just be confusing.

syedsajjadkazmii replied:
Hi @robrap,

I have updated the PR. Could you please have a final look?

robrap replied:
Marking resolved. I'll add updated comments as-needed.

robrap commented on lms/envs/common.py:
```diff
@@ -5041,6 +5044,21 @@ def _make_locale_paths(settings):  # pylint: disable=missing-function-docstring
+# .. toggle_use_cases: temporary
+# .. toggle_creation_date: 2023-12-07
+# .. toggle_target_removal_date: 2024-04-01
+# .. toggle_tickets: https://openedx.atlassian.net/browse/VAN-1694
```
It's really the removal ticket that is more important at this point. I linked the other ticket on removal ticket.
```suggestion
#       email change occurs. This toggle is just being used for rollout.
# .. toggle_use_cases: temporary
# .. toggle_creation_date: 2023-12-07
# .. toggle_target_removal_date: 2024-04-01
# .. toggle_tickets: https://2u-internal.atlassian.net/browse/VAN-1797
```

robrap reviewed (commented):
Thank you. A minor comment (I hope). This is just about ready.

Lastly, once you make any final updates, if you could do one more manual check of masquerading, that would be great.

syedsajjadkazmii committed:
fix: store email in session update

syedsajjadkazmii committed:
fix: add setting for tests

syedsajjadkazmii committed:
fix: fix tests

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -768,6 +768,92 @@ def _get_encrypted_request_headers(request):
+            if request.session.get('email', None) is None:
+                # .. custom_attribute_name: session_with_no_email_found
+                # .. custom_attribute_description: Indicates that user's email was not
+                #      stored in the user's session.
```
Nit: Since we are updating the session.
```suggestion
                #      yet stored in the user's session.
```

robrap reviewed (approved):
Nice. Thank you.

github-actions[bot] committed:
feat: Upgrade Python dependency edx-drf-extensions (#34135)

Commit generated by workflow `openedx/edx-platform/.github/workflows/upgrade-one-python-dependency.yml@refs/heads/master`

Co-authored-by: syedsajjadkazmii <syedsajjadkazmii@users.noreply.github.com>

syedsajjadkazmii committed:
Merge branch 'master' into sajjad/VAN-1694
